### ISSUE TITLE
Is any kind of distributed processing on the roadmap?

### ISSUE THREAD

**Issue by appcrawler** [online-ml/river#438]
I love this framework.  It's simple and has tons of functionality.  I stood up a Flask endpoint in front of the predict_one_method, and it occurred to me: have you considered anything around making it highly available in a distributed way?  For example, I used your reco class.  I am pulling a stream of changes from Kafka.  If I send a user and movie for a recommendation score and have two servers, it needs to either:

1. Distribute the model to both
2. Ensure the user (or some key) is used for sticky persistence to a given server.  However, the assumption is the same score could be returned from either server if the same data is available to both.  We could also have each server pull all the messages from Kafka, so the model is calculated separately on each server.

Any thoughts?

**Comment by MaxHalford** [comment]
Thanks for the kind words, it always means a lot. It's nice to see appreciation for our quest for simplicity :)

I wrote [chantilly](https://github.com/online-ml/chantilly) a little while as a proof of concept of how to deploy an online model. It's cute and simple but doesn't deal with the distributed part that you're mentioning.

I've been giving this some though over the holidays... Currently, I'm thinking of the following architecture:

![sketch](https://user-images.githubusercontent.com/8095957/103723231-e0033d80-4fd1-11eb-92f7-1dbbafc3c8f0.png)

Essentially, I'm considering basing this on something like [AWS Lambda functions](https://aws.amazon.com/fr/lambda/). This would deal with the scaling part with little effort (I think). A background job could take care of processing the training data. Every so often it would update the model in storage.

I like this architecture because the model is only updated once. The model is not duplicated on multiple servers. The version of the model that is being used to make predictions isn't always in sync with the latest training data, but that's okay and reasonable.

Note that we don't have to lock ourselves in to AWS, there's other implementations of serverless functions, such as [Fission](https://platform9.com/blog/fission-the-open-source-alternative-to-aws-lambda/), [OpenFaaS](https://github.com/openfaas/faas), and [Kubeless](https://kubeless.io/).

Ideally, we would provide this is as a standalone tool that is very easy to use, something in the lines of [Cortex](https://www.cortex.dev/). I also want to build something flexible. For instance, we could take care of setting up a Kafka instance on AWS. But you should also be able to plug in your own instance.

Feedback is very much appreciated! I would like to get this crystal clear on paper before implementing it.

**Comment by chappers** [comment]
I'm not a contributor or anything - I've been giving it some thought as well, and the challenges I can see happening are around "how can we be sure the updated model is better than the current one"?

The delayed feedback (to me) does make a lot of sense, as often the labels can only be provided after-the-fact. The only "online" kind of way to update models is in a [Hoeffding Race](https://papers.nips.cc/paper/1993/hash/02a32ad2669e6fe298e607fe7cc0e1a0-Abstract.html) kind of way, where the current "frozen" model, and the new proposed model (with the delayed feedback) are "racing" each other. In this way there is some kind of sane way to provide some (perhaps not all too useful practically) guarentees.

**Comment by MaxHalford** [comment]
Just to be clear: an updated model is a model that has seen a new observation. If so, then  I'm not sure that we should scared that the updated model performs worse than the current model. We should assume that having seen the latest data makes you better at predicting the incoming data. Don't you agree?

**Comment by chappers** [comment]
We can tease this out - I'm more coming from an organisational perspective rather than an academic one. There could be several reasons why we may not want automatically updating models:
*  to guard against potential data poisoning attacks
*  in order to meet regulatory or risk frameworks (in many frameworks if you switch a model off to promote a new model you may need to demonstrate or provide evidence that some kind of testing has been demonstrated so there isn't model degradation)
*  to meet particular service management frameworks or ways of working (e.g. ITIL). Even if a change is pre-approved for a change advisory board (like I presume this one would be, as we are just functionally speaking updating weights) additional governance items or audit items may be required.

From an academic and modelling perspective, I totally agree, it is a fair assumption that a model which has seen updated data ought to perform better than the old model. In many large organisations which are risk adverse, the onus falls on the analyst to demonstrate & document this; even for pre-approved/standard changes in ITIL framework.

**Comment by MaxHalford** [comment]
Thanks for those comments. I'm sure that all the cases you mention are straightforward to implement. Basically they can all be handled if there's a "turn off" button to freeze the model.

**Comment by chappers** [comment]
Correct, in the hoeffding race style, the "original" model would be frozen until the "new" model is deemed ready by <insert metric>

**Comment by jincongho** [comment]
Distributed Serving: 
- export the model and create a wrapper under Cortex seems doable already
- to automate the process we can probably refer to [TFX](https://github.com/tensorflow/tfx), see how to integrate the model validation step

I'm also curious about Distributed Training though, any ideas how to implement such model?

**Comment by MaxHalford** [comment]
There's very little reason for us to help users a model on platforms like Cortex. What we really want to focus is the online training aspect. That's what I'm trying to address in the above chart.
